middlewares/retry.go

```diff
@@ -56,9 +63,12 @@ func (retry *Retry) ServeHTTP(rw http.ResponseWriter, r *http.Request) {
 	}
 }
 
-func isNetworkError(status int) bool {
-	return status == http.StatusBadGateway || status == http.StatusGatewayTimeout
-}
+// NetErrorCtxKey is a custom type that is used as key for the request context.
+// Using an own type we avoid potential collisions with other keys that may have the same string name.
+type NetErrorCtxKey string
```

Comment:
From my learnings about context I remember that the key should be a custom type as you do here, but it should also be non exported, since it's something that should be package internal.
Obviously this is not straightforward here because the server package needs to signal something to the middleware package.
But instead of exporting low level package stuff, should the retry middlewares package maybe export some functions to deal with this and keep the context key type private?
Looks like you already have some kind of function in the server package called: `recordNetError`.
This should probably be moved to the middlewares package.
Also `recordNetError` doesn't need to get passed a request, a context would be sufficient.